### PR TITLE
Add Tuple-based Math.DivRem overloads

### PR DESCRIPTION
**Pull request by EgorBo**
This PR adds `Math.DivRem` APIs approved in https://github.com/dotnet/runtime/issues/42156#issuecomment-729167820

1) Should we hide "legacy" APIs? (see https://github.com/dotnet/runtime/issues/44829) I personally hate them for `out int result` being actually a remainder.
2) Do we use Capitalized names for tuples as return types?
3) I can try to optimize these APIs via single `idiv` since there is a good example by @CarolEidt  https://github.com/dotnet/runtime/pull/37928 (furthermore, we can teach the JIT to recognize a sort of `GT_DIVREM` anywhere now) - do you think it's worth the effort? It was the main reason I decided to add these APIs actually 🙂)

### CODE REVIEW

**Comment by tannergooding** [comment]
> Should we hide "legacy" APIs? (see #44829) I personally hate them for out int result being actually a remainder.

I think we are going to look at an analyzer. I don't know if hiding them is the correct approach, especially for users with muscle memory.

> Do we use Capitalized names for tuples as return types?

Yes.

> I can try to optimize these APIs via single idiv since there is a good example by @CarolEidt #37928 (furthermore, we can teach the JIT to recognize a sort of GT_DIVREM anywhere) - do you think it's worth the effort? (it was the main reason I decided to add these APIs actually 🙂)

I think its probably worth having long term. I'm sure there is a lot of code that doesn't use `DivRem`, either because they don't know it exists or because they don't think about it.

File: src/libraries/System.Private.CoreLib/src/System/Math.cs [hunk 1]
```diff
@@ -285,6 +285,71 @@ internal static ulong DivRem(ulong a, ulong b, out ulong result)
             return div;
         }
 
+        [CLSCompliant(false)]
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:288]: Can we delete the internal methods above and switch callers to use the public ones?

**Comment by jkotas** [comment]
> Do we use Capitalized names for tuples as return types?

Yes. This was discussed extensively in the past and this was the conclusion: https://github.com/dotnet/runtime/issues/27939#issuecomment-531420515

**Comment by davidfowl** [comment]
Someday we'll have better generics 🙃

File: src/libraries/System.Private.CoreLib/src/System/Math.cs (hunk 1 above)
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:274]: Done

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs [hunk 2]
```diff
@@ -1334,7 +1334,8 @@ private static unsafe void UInt32ToNumber(uint value, ref NumberBuffer number)
         {
             while (--digits >= 0 || value != 0)
             {
-                value = Math.DivRem(value, 10, out uint remainder);
+                uint remainder;
+                (value, remainder) = Math.DivRem(value, 10);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1338]: Before merging, can we validate there's no unexpected regression induced by these changes?

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1338]: I tested various patterns including LICM-ability and didn't notice any differences.

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1338]: Actually there was a difference so I removed my changes for https://github.com/dotnet/runtime/pull/45074/commits/60a8d989a8222d144b298b30ae60cc507427e119

Inliner didn't want to inline the new byte version in that code + some redundant moves, will investigate and file an issue.

**Comment by EgorBo** [comment]
A few notes:
`(u)short` and `(s)byte` overloads are above the ALWAYS_INLINE size and don't want to inline (even to a loop body):
```
weight= 10 : state   3 [ ldarg.0 ]
weight= 16 : state   4 [ ldarg.1 ]
weight= 35 : state  79 [ div ]
weight= 25 : state 155 [ conv.u2 ]
weight= 20 : state 199 [ stloc.0 -> ldloc.0 ]
weight= 10 : state   3 [ ldarg.0 ]
weight= 12 : state   7 [ ldloc.0 ]
weight= 16 : state   4 [ ldarg.1 ]
weight= -9 : state  78 [ mul ]
weight=-15 : state  77 [ sub ]
weight= 25 : state 155 [ conv.u2 ]
weight=227 : state 103 [ newobj ]
weight= 19 : state  42 [ ret ]

Inline candidate callsite is in a loop.  Multiplier increased to 3.
calleeNativeSizeEstimate=391
callsiteNativeSizeEstimate=115
benefit multiplier=3
threshold=345
Native estimate for function size exceeds threshold for inlining 39.1 > 34.5 (multiplier = 3)
```
I wonder if we should add `AggressiveInlining` for them, other overloads are fine.

**Comment by jkotas** [comment]
Yes, I think it would be fine to add AggressiveInlining to all these methods.

File: src/libraries/System.Private.CoreLib/src/System/Math.cs [hunk 3]
```diff
@@ -271,18 +271,79 @@ public static long DivRem(long a, long b, out long result)
             return div;
         }
 
-        internal static uint DivRem(uint a, uint b, out uint result)
+        [CLSCompliant(false)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:275]: I think you can also mark these methods as `[NonVersionable]` so they get inlined accross versioning boundaries in R2R.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/StandardFormat.cs [hunk 4]
```diff
@@ -182,7 +182,8 @@ internal int Format(Span<char> destination)
 
                     if (precision >= 10)
                     {
-                        uint div = Math.DivRem(precision, 10, out precision);
+                        uint div;
+                        (div, precision) = Math.DivRem(precision, 10);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/StandardFormat.cs:186]: Is there an advantage using the new overloads here and similar places below?  Does the codegen end up being better?

File: src/libraries/System.Private.CoreLib/src/System/Math.cs [hunk 5]
```diff
@@ -271,18 +271,89 @@ public static long DivRem(long a, long b, out long result)
             return div;
         }
 
-        internal static uint DivRem(uint a, uint b, out uint result)
+        [NonVersionable]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Math.cs:274]: All these public members need XML doc comments.

File: src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs [hunk 6]
```diff
@@ -446,7 +446,7 @@ public static void DivRem(ref BigInteger lhs, ref BigInteger rhs, out BigInteger
 
                 if ((lhsLength == 1) && (rhsLength == 1))
                 {
-                    uint quotient = Math.DivRem(lhs._blocks[0], rhs._blocks[0], out uint remainder);
+                    var (quotient, remainder) = Math.DivRem(lhs._blocks[0], rhs._blocks[0]);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs:449]:
```suggestion
                    (uint quotient, uint remainder) = Math.DivRem(lhs._blocks[0], rhs._blocks[0]);
```

File: src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs [hunk 7]
```diff
@@ -261,7 +261,7 @@ private static ulong ConvertBigIntegerToFloatingPointBits(ref BigInteger value,
                 return AssembleFloatingPointBits(in info, value.ToUInt64(), baseExponent, !hasNonZeroFractionalPart);
             }
 
-            uint topBlockIndex = Math.DivRem(integerBitsOfPrecision, 32, out uint topBlockBits);
+            var (topBlockIndex, topBlockBits) = Math.DivRem(integerBitsOfPrecision, 32);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs:264]:
```suggestion
            (uint topBlockIndex, uint topBlockBits) = Math.DivRem(integerBitsOfPrecision, 32);
```

File: src/libraries/System.Private.CoreLib/src/System/Buffers/StandardFormat.cs (hunk 4 above)
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/StandardFormat.cs:186]: @stephentoub sorry for the delay, didn't notice notifications.

I couldn't find any differences in terms of codegen, this and other changes are made to remove internal DivRem functions

File: src/libraries/System.Private.CoreLib/src/System/Math.cs (hunk 5 above)
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Math.cs:278]: Done.

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**